### ISSUE TITLE
BlockLSTM returns 0 tensor values

### ISSUE THREAD

**Issue by danilojsl** [tensorflow/java#270]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): 
- TensorFlow version (use command below): 2.3.1
- Python version: 3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**
I'm using tensorflow-java as a dependency for my project. I wanted to use *BlockLSTM* feature but the output returns always the same values. e.g. cell state output always has 0 values.

**Describe the expected behavior**
I will expect to see results similar to LSTM Keras layer from Python. In which, each run returns different values
e.g. cell state outputs
run 1: [ 0.12028465,  0.07415504, -0.09205371, -0.14372592,  0.00117318]
run 2: [ 0.07089745, -0.02260131, -0.00052543, -0.19030134,  0.14710784]

**Code to reproduce the issue**
You can check the code I wrote on this public repo:
https://github.com/danilojsl/tensorflow-java-spikes/blob/main/src/main/java/LSTMSpike.java

**Other info / logs**
Warning: Could not load Loader: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Warning: Could not load Pointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Warning: Could not load BytePointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
2021-04-01 12:46:24.406136: I external/org_tensorflow/tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Warning: Could not load IntPointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
Warning: Could not load PointerPointer: java.lang.UnsatisfiedLinkError: no jnijavacpp in java.library.path
2021-04-01 12:46:24.522141: W external/org_tensorflow/tensorflow/core/kernels/rnn/lstm_ops.cc:869] BlockLSTMOp is inefficient when both batch_size and cell_size are odd. You are using: batch_size=1, cell_size=5
Input Gate: [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]
Cell State: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
Forget State: [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]
Output Gate: [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]
Cell Input: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
Cell Output: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
Hidden Output: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

**Comment by karllessard** [comment]
Thanks @danilojsl , can you also share a sample of the Python script that gives you different result?

**Comment by Craigacp** [comment]
I think this is because you've set all the weights to zero, so when it goes through the LSTM equation then everything is multiplied by a zero vector. Try setting some of the weights to random values, or ones.

**Comment by danilojsl** [comment]
Hi, @karllessard. Sure, [here](https://github.com/danilojsl/spike_nn/blob/main/src/tensorflow/model_subclassing_lstm.py
) is the Python script.

I assume LSTM Layer in Keras does many more things under the hood. So, I will follow @Craigacp suggestion to initialize all weights with `glorot_uniform` and let you know how it goes.

Thanks for your fast replies

**Comment by danilojsl** [comment]
I was able to get reasonable values as `BlockLSTM` output. Since I require to work with *TensorFlow Java Version 0.2.0* I used `TruncatedNormal` as an initializer for the weights.

Now, I'm a little confused because there is also a `BlockLSTMGrad` class that according to the documentation should be used in conjunction with `BlockLSTM`.
As much as I understand `BlockLSTM` output will be the input of `BlockLSTMGrad`, but it requires two arguments that I don't know how to get from:
```
csGrad – The current gradient of cs.
hGrad – The gradient of h vector
```

Could you please guide me on how I can compute those values?

Do I need to use both `BlockLSTM` and `BlockLSTMGrad` to have a behavior similar to Keras LSTM?

**Comment by Craigacp** [comment]
I don't think that operation is meant to be used directly, it's added to graphs when you apply gradients to them via an `Optimizer`.

**Comment by danilojsl** [comment]
So, only with `BlockLSTM` will be enough to have a behavior similar to Keras LSTM as in [this](https://github.com/danilojsl/spike_nn/blob/main/src/tensorflow/model_subclassing_lstm.py) Python script?

**Comment by Craigacp** [comment]
Probably not, the LSTM in Keras is quite complicated and only calls the op we expose at the very end. I've not traced through the computation properly to see what preprocessing we'd need to add to make an equivalent of the Keras LSTM layer. We're still working on building up higher level functionality like that in the Java API, and recurrent nets haven't been started yet.

**Comment by danilojsl** [comment]
I see, thanks for the help anyways. It would be amazing to have that feature.
I'm a contributor to [Spark NLP](https://github.com/JohnSnowLabs/spark-nlp) API and we plan to port as much as possible our TF operations from Python to Java/Scala.
We want to use this tensorflow-java library looking forward to having our TF models entirely created with Java. Starting with our [release 3.0.0](https://github.com/JohnSnowLabs/spark-nlp/releases/tag/3.0.0), we added tensorflow-java as our dependency.
I checked that there is a Feature Request category when creating a new issue. It would be ok if I open a new feature request for this LSTM-RNN layer?

**Comment by karllessard** [comment]
Yes sure please do, that would be something interesting to look at!

CCing @JimClarke5 which is our expert in porting Keras code to Java

**Comment by mullerhai** [comment]
@danilojsl   HI ,I want to know does your team had make the  LSTM Layer to real with tensorflow-java in spark_nlp project. I also  develop one project  like  keras-scala, want to make  the rnn  lstm  gru  transformer  Layer to real  use tensorflow-java

**Comment by danilojsl** [comment]
Hi @mullerhai, no in spark-nlp we haven't implemented that feature and won't do it in the foreseeable future. We are still waiting for contributors of tensorflow-java to make this available.
